wiki/static/wiki/js/urlify.js

```diff
@@ -0,0 +1,185 @@
+        'Ð': 'Dz', 'Ä': 'Dj'
+    };
+    var AZERBAIJANI_MAP = {
+        'Ã§': 'c', 'É™': 'e', 'ÄŸ': 'g', 'Ä±': 'i', 'Ã¶': 'o', 'ÅŸ': 's', 'Ã¼': 'u',
+        'Ã‡': 'C', 'Æ': 'E', 'Äž': 'G', 'Ä°': 'I', 'Ã–': 'O', 'Åž': 'S', 'Ãœ': 'U'
+    };
+    var GEORGIAN_MAP = {
+        'áƒ': 'a', 'áƒ‘': 'b', 'áƒ’': 'g', 'áƒ“': 'd', 'áƒ”': 'e', 'áƒ•': 'v', 'áƒ–': 'z',
+        'áƒ—': 't', 'áƒ˜': 'i', 'áƒ™': 'k', 'áƒš': 'l', 'áƒ›': 'm', 'áƒœ': 'n', 'áƒ': 'o',
+        'áƒž': 'p', 'áƒŸ': 'j', 'áƒ ': 'r', 'áƒ¡': 's', 'áƒ¢': 't', 'áƒ£': 'u', 'áƒ¤': 'f',
+        'áƒ¥': 'q', 'áƒ¦': 'g', 'áƒ§': 'y', 'áƒ¨': 'sh', 'áƒ©': 'ch', 'áƒª': 'c', 'áƒ«': 'dz',
+        'áƒ¬': 'w', 'áƒ­': 'ch', 'áƒ®': 'x', 'áƒ¯': 'j', 'áƒ°': 'h'
+    };
+
+    var ALL_DOWNCODE_MAPS = [
+        LATIN_MAP,
+        LATIN_SYMBOLS_MAP,
+        GREEK_MAP,
+        TURKISH_MAP,
+        ROMANIAN_MAP,
+        RUSSIAN_MAP,
+        UKRAINIAN_MAP,
+        CZECH_MAP,
+        POLISH_MAP,
+        LATVIAN_MAP,
+        ARABIC_MAP,
+        LITHUANIAN_MAP,
+        SERBIAN_MAP,
+        AZERBAIJANI_MAP,
+        GEORGIAN_MAP
+    ];
+
+    var Downcoder = {
+        'Initialize': function() {
+            if (Downcoder.map) {  // already made
+                return;
+            }
+            Downcoder.map = {};
+            Downcoder.chars = [];
+            for (var i = 0; i < ALL_DOWNCODE_MAPS.length; i++) {
+                var lookup = ALL_DOWNCODE_MAPS[i];
+                for (var c in lookup) {
+                    if (lookup.hasOwnProperty(c)) {
+                        Downcoder.map[c] = lookup[c];
+                    }
+                }
+            }
+            for (var k in Downcoder.map) {
+                if (Downcoder.map.hasOwnProperty(k)) {
+                    Downcoder.chars.push(k);
+                }
+            }
+            Downcoder.regex = new RegExp(Downcoder.chars.join('|'), 'g');
+        }
+    };
+
+    function downcode(slug) {
+        Downcoder.Initialize();
+        return slug.replace(Downcoder.regex, function(m) {
+            return Downcoder.map[m];
+        });
+    }
+
+
+    function URLify(s, num_chars, allowUnicode) {
+        // changes, e.g., "Petty theft" to "petty-theft"
+        // remove all these words from the string before urlifying
+        if (!allowUnicode) {
+            s = downcode(s);
+        }
+
+        /*
+        THIS IS THE PATCHED PART! We removed all the words and thus preserve
+        any "and", "the", "an" etc. in the slug.
+        */
+        var removelist = [];
+
+        var r = new RegExp('\\b(' + removelist.join('|') + ')\\b', 'gi');
+        s = s.replace(r, '');
+        // if downcode doesn't hit, the char will be stripped here
+        if (allowUnicode) {
+            // Keep Unicode letters including both lowercase and uppercase
+            // characters, whitespace, and dash; remove other characters.
+            s = XRegExp.replace(s, XRegExp('[^-_\\p{L}\\p{N}\\s]', 'g'), '');
```

Comment:
Where is `XRegExp` defined? I'm trying to use this code in a custom form and am getting an error.

Reply:
Oh I see, I need to include the script from /static/admin/js/vendor/xregexp/xregexp.min.js to use `allowUnicode=true`.